lib/utils/hostname.go

```diff
@@ -0,0 +1,104 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"golang.org/x/exp/slices"
+)
+
+func isFQDN(hostname string) bool {
+	return strings.Contains(hostname, ".")
+}
+
+func isWildcard(hostname string) bool {
+	return strings.HasPrefix(hostname, "*.")
+}
+
+func wildcardFromHostname(hostname string) string {
+	if isWildcard(hostname) {
+		return hostname
+	}
+	return fmt.Sprintf("*.%s", strings.Join(strings.Split(hostname, ".")[1:], "."))
+}
+
+// AddHostToHostList adds the provided hostname to the hostList, then processes the hostList to add wildcards
+// if appropriate. it returns a processed, sorted hostList.
+//
+// for every host in the list:
+//   - all hostnames are converted to their wildcard form (test.example.com -> *.example.com) and grouped together.
+//   - if we only have a single hostname which would be matched by a given wildcard, that hostname is added
+//     to the hostList verbatim to prevent matching against too many hosts with a wildcard.
+//   - if a wildcard matches multiple hostnames, the wildcard will be added to the hostList instead.
+//
+// this is an effort to keep the length of hostList as short as possible for efficiency and tidiness.
+func AddHostToHostList(hostList []string, hostname string) []string {
+	// add the incoming hostname to the hostList before we sort and process it
+	hostList = append(hostList, hostname)
+
+	hostMap := make(map[string][]string)
+	// iterate over the full hostList
+	// if the element is a wildcard, add it to the list of wildcards
+	// if the element is not a wildcard, convert it to a wildcard and add any hostnames it matches to a map
+	for _, element := range hostList {
+		// FQDN-based hosts are grouped under a wildcard key
+		if isFQDN(element) {
+			wildcard := wildcardFromHostname(element)
+			if !slices.Contains(hostMap[wildcard], element) {
+				hostMap[wildcard] = append(hostMap[wildcard], element)
+			}
+		} else {
+			// any non-wildcard hosts go into the "extra" key and will be processed separately at the end
+			hostMap["extra"] = append(hostMap["extra"], element)
+		}
+	}
+
+	// iterate over the map, look for all wildcard keys with more than one hostname matching.
+	// for each match, add the wildcard to the hostList.
+	var outputHostList []string
+	for key, matchingHostnames := range hostMap {
+		// add all non-wildcard matches separately
+		if key == "extra" {
+			for _, hostname := range matchingHostnames {
+				if !slices.Contains(outputHostList, hostname) {
+					outputHostList = append(outputHostList, hostname)
```

Comment:
How many hostnames do we expect in the general case? What about at the high end?

I'm not confident this will perform great with a large number of hostnames (it's essentially 3 nested iterations), but I don't have enough context to know if that will matter or not.

Reply:
I agree that the performance will not be stellar with large numbers of hostnames.

I'd say that _generally_, this is likely to involve a low number of hostnames (~5-20). What I'd expect to happen in most cases is that people will run `tsh puttyconfig` for just a few hosts that they access regularly, which will keep the list short - or they will add new hosts as and when they need them, rather than in large batches.

PuTTY's saved session UI is a short, narrow box with limited search functionality which makes it not particularly conducive to adding huge numbers of hosts and finding the one you need. It's quite self-limiting in this regard:

![image](https://user-images.githubusercontent.com/897821/236536073-e2405841-b1f7-416e-8173-4c6db4e25cce.png)

If someone were to loop `tsh puttyconfig` over 500 hostnames in the same domain, this would be caught by the wildcard which is added as soon as the number of hosts for a given domain is >1 - where it will check that `host500.example.com` matches `*.example.com` pretty early on and then just return that. That should be fairly efficient. Even with 10-20 different domains/subdomains this would perform fine.

If someone were to loop over 1000 individual hostnames with no `.` (so unable to be matched as any kind of FQDN) then I'd expect this loop to get progressively slower with each iteration. However, I'd speculate that on any modern machine, it would still be fairly fast even with ~1000 hostnames (which I'd probably say is the absolute upper bound for the number of hosts anyone would add, which would be utterly unusable at that point from a UI perspective in PuTTY anyway)

TL;DR It's probably fine.